### PR TITLE
fix(describe-pp): try more candidates for pped dialect sources

### PR DESCRIPTION
**Pull request by anmonteiro**
- I was trying out this feature, and I noticed that dune doesn't always produce the same files for preprocessed dialect sources.
- this change looks for both `foo.pp.re.ml` and `foo.re.pp.ml` (which can happen based on whether preprocessing is `(action ..)` or `(pps ..)`

### CODE REVIEW

File: bin/describe/describe_pp.ml [hunk 1]
```diff
@@ -54,16 +70,20 @@ let get_pped_file super_context file =
     then User_error.raise [ Pp.textf "No file given." ]
     else Path.of_string file |> in_build_dir |> Path.build
   in
-  let pp_file = file_in_build_dir |> Path.map_extension ~f:(fun ext -> ".pp" ^ ext) in
-  Build_system.file_exists pp_file
+  let* project = Source_tree.root () >>| Source_tree.Dir.project in
+  let dialects = Dune_project.dialects project in
+  let ml_kind, pp_files_to_check = files_for_source file_in_build_dir dialects in
+  Memo.all
+    (List.map
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply bin/describe/describe_pp.ml]: good point, changed to `Memo.parallel_map`.

File: bin/describe/describe_pp.ml [hunk 2]
```diff
@@ -19,28 +19,44 @@ let pp_with_ocamlc env ~ocamlc pp_file dump_file =
...
-      ~ext:
-        (match kind with
-         | Intf -> ".cmi.dump"
-         | Impl -> ".cmo.dump")
+    | None ->
+      (* No `.ml` suffix for this dialect. this is already an `.ml` file.
+         The extension for the pped file is `.pp.ml` *)
+      [ pp_file_result_base ]
+    | Some suffix ->
+      (* If there's an `.ml` suffix for this dialect, append it to the source.
```
Comment (rgrinberg) [inline bin/describe/describe_pp.ml:49]: I find this check to be unnecessary. Why do we need to guess if a file exists or not if we are the ones setting up the rules? Can't we just know what the source should be from `Module.t`?

Comment (anmonteiro) [reply bin/describe/describe_pp.ml:49]: I guess so. What's the best way to get a `Module.t` from the source path that the user passes in the CLI to describe pp?

**Comment by anmonteiro** [comment]
In the latest version:

- derive `Module.t` from the file passed via the CLI
- change the `staged_pps` check to use the new code

**Review by rgrinberg — APPROVED** [review]

File: bin/describe/describe_pp.ml [hunk 3]
```diff
@@ -45,72 +48,140 @@ let print_pped_file sctx file pp_file =
...
-       User_error.raise
-         [ Pp.textf "%s does not exist" (Path.to_string_maybe_quoted file_in_build_dir) ])
+  let* ml_kind =
+    let+ _, ml_kind = dialect_and_ml_kind file in
+    ml_kind
+  in
+  let* module_for_file = module_for_file ~sctx:super_context ~ml_kind file_in_build_dir in
+  let build_fallback =
+    Memo.lazy_ (fun () ->
+      Build_system.file_exists file_in_build_dir
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply bin/describe/describe_pp.ml]: I suppose the fallback isn't actually needed now that we can precisely find the corresponding `Module.t` for the file argument. I removed it in favor or raising the error in both places.

File: bin/describe/describe_pp.ml [hunk 4]
```diff
@@ -45,72 +48,140 @@ let print_pped_file sctx file pp_file =
...
+            Some (Exe { first_exe = snd (List.hd exes.names) }, exes.buildable.preprocess)
+          | Melange_stanzas.Emit.T mel ->
+            Some (Melange { target = mel.target }, mel.preprocess)
+          | _ -> None
+        in
+        match for_and_preprocess with
+        | None -> Memo.return None
+        | Some (for_, preprocess) ->
+          let+ modules =
+            Dune_rules.Dir_contents.get sctx ~dir
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply bin/describe/describe_pp.ml:114]: Interesting. I'll look into seeing whether that's possible.

**Pull request merged**